kvbc/test/categorization/versioned_kv_category_unit_test.cpp

```diff
@@ -0,0 +1,625 @@
+    ASSERT_EQ(proof->value, "v3");
+    ASSERT_EQ(proof->key_value_index, 4);  // indexing starts at 0
+    ASSERT_THAT(proof->calculateRootHash(),
+                ContainerEq(Hash{0xac, 0xc0, 0x73, 0x12, 0xb8, 0x6d, 0xae, 0xc7, 0xd0, 0xb8, 0xb5,
+                                 0x3b, 0xd6, 0x7e, 0x49, 0xd7, 0xc5, 0xc1, 0x67, 0x7b, 0xf1, 0x74,
+                                 0x90, 0x49, 0x62, 0xa8, 0x36, 0xeb, 0x9d, 0xb4, 0xa8, 0xcf}));
+  }
+
+  // Invalid block ID.
+  {
+    ASSERT_FALSE(cat.getProof(invalid_block, "k1", out));
+    ASSERT_FALSE(cat.getProof(invalid_block, "k2", out));
+    ASSERT_FALSE(cat.getProof(invalid_block, "k3", out));
+  }
+
+  // Non-existing key.
+  { ASSERT_FALSE(cat.getProof(block1, "non-existing", out)); }
+}
+
+TEST_F(versioned_kv_category, delete_key) {
+  const auto stale_on_update = false;
+
+  // Update key "k" to "v1" in block 1.
+  {
+    auto in = VersionedInput{};
+    in.calculate_root_hash = true;
+    in.kv["k"] = ValueWithFlags{"v1", stale_on_update};
+    add(1, std::move(in));
+  }
+
+  // Delete key "k" in block 2.
+  {
+    auto in = VersionedInput{};
+    in.calculate_root_hash = true;
+    in.deletes.push_back("k");
+    const auto out = add(2, std::move(in));
+    ASSERT_EQ(out.keys.size(), 1);
+    ASSERT_EQ(out.keys.cbegin()->first, "k");
+    ASSERT_TRUE(out.keys.cbegin()->second.deleted);
+  }
+
+  // Key "k" is still available at block 1 via get().
+  {
+    const auto value = cat.get("k", 1);
+    ASSERT_TRUE(value);
+    ASSERT_EQ(asVersioned(value).block_id, 1);
+    ASSERT_EQ(asVersioned(value).data, "v1");
+  }
+
+  // Key "k" is still available at block 1 via multiGet().
+  {
+    auto values = std::vector<std::optional<categorization::Value>>{};
+    cat.multiGet({"k"}, {1}, values);
+    const auto expected = std::vector<std::optional<categorization::Value>>{VersionedValue{{1, "v1"}}};
+    ASSERT_EQ(values, expected);
+  }
+
+  // Key "k" is no longer available for blocks >= 2 via get().
+  {
+    ASSERT_FALSE(cat.get("k", 2));
+    ASSERT_FALSE(cat.get("k", 3));
+  }
+
+  // Key "k" is no longer available for blocks >= 2 via multiGet().
+  {
+    auto values = std::vector<std::optional<categorization::Value>>{};
+    cat.multiGet({"k", "k"}, {2, 3}, values);
+    const auto expected = std::vector<std::optional<categorization::Value>>{std::nullopt, std::nullopt};
+    ASSERT_EQ(values, expected);
+  }
+
+  // Key "k" has no latest value via getLatest().
+  { ASSERT_FALSE(cat.getLatest("k")); }
+
+  {
+    // Key "k" has no latest value via multiGetLatest().
+    auto values = std::vector<std::optional<categorization::Value>>{};
+    cat.multiGetLatest({"k"}, values);
+    const auto expected = std::vector<std::optional<categorization::Value>>{std::nullopt};
+    ASSERT_EQ(values, expected);
+  }
+
+  // Key "k" has a latest version of 2 via getLatestVersion() - the version it was deleted at.
+  {
+    const auto latest_version = cat.getLatestVersion("k");
```

Comment:
I find this behavior confusing when viewed in isolation. In the BlockMerkle code, `std::nullopt` is returned instead to indicate that the key was deleted. Here that doesn't happen, and the user has to read the key to discover that.

As this is related to the rest of the code, I'm fine leaving it for now, but we may want to unify the behavior among the categories one way or the other. 

I'm wondering whether this is the correct behavior and not the behavior in BlockMerkle, because of how conflict detection must be done with pre-execution. We may need to know that the key was deleted at a given block id. I guess I'll figure that out as I work on the deletion part of the BlockMerkle category.

Reply:
Will use the approach taken in https://github.com/vmware/concord-bft/pull/1120.